x/tally/keeper/payout.go

```diff
@@ -35,27 +58,110 @@ func (k Keeper) CalculateCommitterPayouts(ctx sdk.Context, req types.Request, ga
+	adjGasUsed := min(gasReport, execGasLimit/uint64(replicationFactor))
+	payoutAmt := gasPrice.Mul(math.NewIntFromUint64(adjGasUsed))
+	burnAmt := math.ZeroInt()
+	if reducedPayout {
+		burnRatio := math.LegacyNewDecWithPrec(2, 1) // burn 20%
+		burnAmt := burnRatio.MulInt(payoutAmt).TruncateInt()
+		payoutAmt = payoutAmt.Sub(burnAmt)
+	}
+
+	totalBurn := math.ZeroInt()
+	dists := make([]types.Distribution, len(executors))
+	for i, executor := range executors {
+		if burnAmt.IsPositive() {
+			totalBurn = totalBurn.Add(burnAmt)
+		}
+		dists[i] = types.NewExecutorReward(executor, payoutAmt)
+
+		ctx.EventManager().EmitEvent(sdk.NewEvent(types.EventTypeExecutorRewardUniform,
+			sdk.NewAttribute(types.AttributeExecutor, executor),
+			sdk.NewAttribute(sdk.AttributeKeyAmount, payoutAmt.String()),
+		))
+	}
+	return dists, adjGasUsed * uint64(replicationFactor), totalBurn
+}
+
+// CalculateDivergentPayouts calculates payouts for the executors given their
+// divergent gas reports. It also returns the total execution gas consumption.
+// It assumes that the i-th executor is the one who revealed the i-th reveal.
+func (k Keeper) CalculateDivergentPayouts(ctx sdk.Context, executors []string, gasReports []uint64, execGasLimit uint64, replicationFactor uint16, gasPrice math.Int, reducedPayout bool) ([]types.Distribution, uint64, math.Int) {
+	adjGasUsed := make([]uint64, len(gasReports))
+	var lowestGasUsed uint64
+	var lowestReporterIndex int
+	for i, gasReport := range gasReports {
+		adjGasUsed[i] = min(gasReport, execGasLimit/uint64(replicationFactor))
+		if i == 0 || adjGasUsed[i] < lowestGasUsed {
+			lowestReporterIndex = i
+			lowestGasUsed = adjGasUsed[i]
+		}
+	}
+	medianGasUsed := median(adjGasUsed)
+	totalGasUsed := medianGasUsed*uint64(replicationFactor-1) + min(lowestGasUsed*2, medianGasUsed)
+	totalShares := medianGasUsed*uint64(replicationFactor-1) + lowestGasUsed*2
+	lowestPayout := gasPrice.Mul(math.NewIntFromUint64(lowestGasUsed * 2 * totalGasUsed / totalShares))
+	normalPayout := gasPrice.Mul(math.NewIntFromUint64(medianGasUsed * totalGasUsed / totalShares))
+
+	burnRatio := math.LegacyNewDecWithPrec(2, 1) // burn 20% in case of reduced payout
+
+	totalBurn := math.ZeroInt()
+	dists := make([]types.Distribution, len(executors))
+	for i, executor := range executors {
+		payout := normalPayout
+		if i == lowestReporterIndex {
+			payout = lowestPayout
+		}
+		if reducedPayout {
+			burnAmt := burnRatio.MulInt(payout).TruncateInt()
+			totalBurn = totalBurn.Add(burnAmt)
+			payout = payout.Sub(burnAmt)
 		}
+		dists[i] = types.NewExecutorReward(executor, payout)
+
+		ctx.EventManager().EmitEvent(sdk.NewEvent(types.EventTypeExecutorRewardUniform,
+			sdk.NewAttribute(types.AttributeExecutor, executor),
+			sdk.NewAttribute(sdk.AttributeKeyAmount, payout.String()),
+		))
```

Comment:
In terms of events I think we're only really interested in 2 values: `medianGas` and `lowestReporter` (which can be empty for uniform scenarios). This will allow us to simple add those attributes to the DR completion event and simplify indexing in that regard. The adjusted values after the shares calculation we will receive from the contract when it sends the actual SEDA amounts to the identities.

Reply:
Please take a look at the function `DistributionsFromGasCalculation()` and let me know what you think. If we want those values, we can add those values to the `GasCalculation` struct and emit them as events later, too.

Reply:
I think I can determine those values when indexing the event, so if it means we don't have to make big changes we can leave the implementation as is on the chain side.